File: solutions/go/phone-number/1/phone_number.go
package phonenumber

import (
	"errors"
	"strings"
)

// Number returns the valid phone number in the given string if found.
func Number(phoneNumber string) (string, error) {
	var sb strings.Builder

	for _, ch := range phoneNumber {
		if ch >= '0' && ch <= '9' {
			sb.WriteRune(ch)
		}
	}

	numberString := sb.String()

	if sb.Len() < 10 || sb.Len() > 11 {
		return "", errors.New("invalid number")
	} else if sb.Len() == 11 {
		if numberString[0] != '1' {
			return "", errors.New("invalid number")
		}
		numberString = sb.String()[1:]
	}

	if numberString[0] < '2' || numberString[3] < '2' {
		return "", errors.New("invalid number")
	}

	return numberString, nil
}

// AreaCode returns only the area code of a valid phone number.
func AreaCode(phoneNumber string) (string, error) {
	number, err := Number(phoneNumber)

	if err != nil {
		return "", err
	}
	return number[:3], nil
}

// Format returns a formatted string of a phone number.
func Format(phoneNumber string) (string, error) {
	number, err := Number(phoneNumber)

	if err != nil {
		return "", err
	}
	return "(" + number[:3] + ") " + number[3:6] + "-" + number[6:], nil
}

File: solutions/c/phone-number/1/phone_number.c
#include "phone_number.h"

#include <stdlib.h>
#include <string.h>
#include <stdio.h>

char *phone_number_clean(const char *input) {
    char* res = malloc(sizeof(char) * (PHONE_NUMBER_LEN + 1));
    res[PHONE_NUMBER_LEN] = '\0';

    int len = strlen(input);
    int index = PHONE_NUMBER_LEN - 1;

    // Find the first digit in the phone number, which is the area code
    int i = 0;
    while (input[i] < '0' || input[i] > '9') {
        i++;
    }
    int area_code = input[i];

    // Iterate in reverse and count digits found
    // Also update final result when digit is found
    int digits = 0;
    for (int i = len - 1; i >= 0; i--) {
        if (input[i] >= '0' && input[i] <= '9') {
            if (index >= 0) {
                // Check if the area/exchance code are valid
                if (index == 0 && input[i] < '2') {
                    memset(res, '0', 10);
                    return res;
                } else if (index == 3 && input[i] < '2') {
                    memset(res, '0', 10);
                    return res;
                }
                res[index] = input[i];
                index--;
            }
            digits++;
        }
    }

    // Determine if phone number had correct amount of digits
    // Check the area code if it was 11 digits
    if ((digits != 10 && digits != 11) || (digits == 11 && area_code != '1')) {
        memset(res, '0', 10);
    }

    return res;
}